src/System.Text.Json/src/System/Text/Json/Document/JsonDocument.Parse.cs

```diff
@@ -147,14 +147,19 @@ public static JsonDocument Parse(string json, JsonReaderOptions readerOptions =
             JsonReaderOptions readerOptions,
             byte[] extraRentedBytes)
         {
+            if (readerOptions.MaxDepth == 0)
+            {
+                readerOptions.MaxDepth = JsonReaderState.DefaultMaxDepth;
+            }
+
             ReadOnlySpan<byte> utf8JsonSpan = utf8Json.Span;
             Utf8JsonReader reader = new Utf8JsonReader(
                 utf8JsonSpan,
                 isFinalBlock: true,
                 new JsonReaderState(options: readerOptions));
 
             var database = new MetadataDb(utf8Json.Length);
-            var stack = new StackRowStack(JsonReaderState.DefaultMaxDepth * StackRow.Size);
+            var stack = new StackRowStack(checked (readerOptions.MaxDepth * StackRow.Size));
```

Comment:
Really?  Couldn't someone have previously used int.MaxValue as the MaxDepth to say "I don't want a max"?

Reply:
Previously someone could have used int.MaxValue because we never used the maxdepth option within the JsonDocument at all (and just used the default). Should we continue to do that (or put a cap of 64 or some other size)? I don't mind making the change but we should agree with what we want the behavior of the JsonDocument to be.

cc @bartonjs

Reply:
StackRowStack has grow-code, that's just the initial size.  I'd leave this allocation as the default.